cmd/env.go

```diff
@@ -75,38 +83,61 @@ uniquely identifies the cluster.
 
 environments/
   default/           [Default generated environment]
-    k.libsonnet
-    k8s.libsonnet
-    swagger.json
-    spec.json
-  us-west/
-    staging/         [Example of user-generated env]
+    .metadata/
       k.libsonnet
       k8s.libsonnet
       swagger.json
-      spec.json      [This will contain the uri of the environment]`,
+    spec.json
+    default.jsonnet
+  us-west/
+    staging/         [Example of user-generated env]
+      .metadata/
+        k.libsonnet
+        k8s.libsonnet
+        swagger.json
+      spec.json      [This will contain the uri of the environment and other environment metadata]
+      staging.jsonnet`,
 	RunE: func(cmd *cobra.Command, args []string) error {
 		return fmt.Errorf("Command 'env' requires a subcommand\n\n%s", cmd.UsageString())
 	},
 }
 
 var envAddCmd = &cobra.Command{
-	Use:   "add <env-name> <env-uri>",
+	Use:   "add <env-name>",
 	Short: "Add a new environment to a ksonnet project",
 	RunE: func(cmd *cobra.Command, args []string) error {
 		flags := cmd.Flags()
-		if len(args) != 2 {
-			return fmt.Errorf("'env add' takes two arguments, the name and the uri of the environment, respectively")
+		if len(args) != 1 {
+			return fmt.Errorf("'env add' takes exactly one argument, which is the name of the environment")
 		}
 
-		envName := args[0]
-		envURI := args[1]
+		name := args[0]
 
-		envNamespace, err := flags.GetString(flagEnvNamespace)
+		uri, namespace, context, err := commonEnvFlags(flags)
 		if err != nil {
 			return err
 		}
 
+		if len(uri) == 0 {
```

Comment:
This is an interesting bit of code because the mutual exclusion of the `--uri` and `--context` flags is baked into the semantics. I'd personally prefer that we hide the call to `server` behind `commonEnvFlags`, and guaranteeing that we always return a valid triple of {uri, ns, ctx}. So, basically, all the logic of this conditional would probably be in `commonEnvFlags`.

Thoughts?

Reply:
I had the call to `server()` hidden in `commonEnvFlags` originally, but had to pull it out because of some fundamental difference in behaviors between `env set` and env add`.

i.e.,
if the `--context` and `--uri` flags are not provided for `env add`, we want to use the default context, pulled from kubeconfig. However, for `env set`, it just means the user doesn't want to update their environment's uri-related metadata. It becomes not obvious when `commonEnvFlags` should call `server()` since it is command-agnostic.

Let me know if you see an alternative.

Reply:
Ok, I think this is not a thing I'm prepared to hold up the review for.